src/index.js

```diff
@@ -292,6 +293,54 @@ class Libp2p extends EventEmitter {
     return connection
   }
 
+  /**
+   * Get peer advertising multiaddrs by concating the addresses used
+   * by transports to listen with the announce addresses.
+   * Duplicated addresses and noAnnounce addresses are filtered out.
+   * This takes into account random ports on matching noAnnounce addresses.
+   * @return {Array<Multiaddr>}
+   */
+  getAdvertisingMultiaddrs () {
+    // Filter noAnnounce multiaddrs
+    const filterMa = this.addressManager.getNoAnnounceMultiaddrs()
+
+    // Special filter for noAnnounce addresses using a random port
+    // eg /ip4/0.0.0.0/tcp/0 => /ip4/192.168.1.0/tcp/58751
+    const filterSpecial = filterMa
+      .map((ma) => ({
+        protos: ma.protos(),
+        ...ma.toOptions()
+      }))
+      .filter((op) => op.port === 0)
+
+    // Create advertising list
+    return this.transportManager.getAddrs()
+      .concat(this.addressManager.getAnnounceMultiaddrs())
+      .filter((ma, index, array) => {
+        // Filter out if repeated
+        if (array.findIndex((otherMa) => otherMa.equals(ma)) !== index) {
+          return false
+        }
+
+        // Filter out if in noAnnounceMultiaddrs
+        if (filterMa.find((fm) => fm.equals(ma))) {
+          return false
+        }
+
+        // Filter out if in the special filter
```

Comment:
I would remove support for this right now and we can look at doing a followup PR for this with verbose testing. I think this logic isn't quite right and non TCP addresses might behave unexpectedly here.